File: Src/MusicPlayerBackend/Services/UserResolver.cs
using Microsoft.AspNetCore.Identity;
using MusicPlayerBackend.Data.Entities;

namespace MusicPlayerBackend.Services;

public interface IUserProvider
{
    Task<User> GetUserAsync();
    Task<User?> GetUserOrDefaultAsync();
    Task<Guid> GetUserIdAsync();
    Task<Guid?> GetUserIdOrDefaultAsync();
}

public sealed class UserProvider(IHttpContextAccessor httpContextAccessor, UserManager<User> userManager) : IUserProvider
{
    public async Task<User> GetUserAsync()
    {
        var claimsPrincipal = httpContextAccessor.HttpContext!.User;
        var user = await userManager.GetUserAsync(claimsPrincipal);

        if (user == default)
            throw new NullReferenceException($"Can't find user. Claims: {claimsPrincipal}");

        return user;
    }

    public async Task<User?> GetUserOrDefaultAsync()
    {
        var claimsPrincipal = httpContextAccessor.HttpContext?.User;
        if (claimsPrincipal == default)
            return default;

        var user = await userManager.GetUserAsync(claimsPrincipal);
        if (user == default)
            throw new NullReferenceException($"Can't find user. Claims: {claimsPrincipal}");

        return user;
    }

    public async Task<Guid> GetUserIdAsync()
    {
        var claimsPrincipal = httpContextAccessor.HttpContext!.User;
        var user = await userManager.GetUserAsync(claimsPrincipal);

        if (user == default)
            throw new NullReferenceException($"Can't find user. Claims: {claimsPrincipal}");

        return user.Id;
    }

    public async Task<Guid?> GetUserIdOrDefaultAsync()
    {
        var claimsPrincipal = httpContextAccessor.HttpContext?.User;
        if (claimsPrincipal == default)
            return default;

        var user = await userManager.GetUserAsync(claimsPrincipal);

        if (user == default)
            throw new NullReferenceException($"Can't find user. Claims: {claimsPrincipal}");

        return user.Id;
    }
}

File: Src/MusicPlayerBackend.Host/UserResolver.fs
namespace MusicPlayerBackend.Host

open System
open System.Threading.Tasks
open Microsoft.AspNetCore.Http
open Microsoft.AspNetCore.Identity

open MusicPlayerBackend.Data.Entities

type IUserProvider =
    abstract GetUser: unit -> Task<User>
    abstract TryGetUser: unit -> Task<User option>
    abstract GetUserId: unit -> Task<Guid>
    abstract TryGetUserId: unit -> Task<Guid option>

[<Sealed>]
type UserProvider(httpContextAccessor: IHttpContextAccessor, userManager: UserManager<User>) as this =
    let self = (this :> IUserProvider)

    interface IUserProvider with
        member this.GetUser() = task {
            return! userManager.GetUserAsync(httpContextAccessor.HttpContext.User)
        }

        member this.GetUserId() = task {
            let! user = self.GetUser()
            return user.Id
        }

        member this.TryGetUserId() = task {
            let! user = self.TryGetUser()
            return user |> Option.map _.Id
        }

        member this.TryGetUser() = task {
            match Option.ofObj httpContextAccessor.HttpContext.User with
            | None ->
                return None
            | Some user ->
                let! user = userManager.GetUserAsync(user)
                return Some user
        }
